### PR TITLE
Initialize msg passing folder for development

### PR DESCRIPTION
**Pull request by hexu33**
Issue Number: None

Objective of pull request: Initialize message passing infrastructure

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?
None folder for refactoringing msg passing.

## What is the new behavior?
- Add cpp_message_infrastructure folder and initialize for developing the project

## Does this introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

File: src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_channel.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+#ifndef ABSTACT_CHANNEL_H_
+#define ABSTACT_CHANNEL_H_
+
+#include "abstract_port.h"
+#include "utils.h"
+
+class AbstractChannel {
+ public:
+  AbstractSendPort *src_port_ = NULL;
```
Comment (killight98) [inline src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_channel.h:13]: replace `AbstractSendPort *` by following define.
```cpp
using SendPortPtr = AbstractSendPort *;
```
Pls use smart pointers (unique_ptr / shared_ptr).

File: src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_channel.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  AbstractSendPort *dst_port_ = NULL;
```
Comment (killight98) [inline src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_channel.h:14]: is it `AbstractRecvPort`?

File: src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_port_implementation.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  int Start();
+  int Join();
+  std::vector<int> GetShape();
+  std::vector<AbstractPort> GetPorts();
+
+  DataType dtype_;
+  std::vector<int> shape_;
+  size_t size_;
+  ProcessModel process_model_;
+  std::vector<AbstractPort> ports_;
```
Comment (killight98) [inline src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_port_implementation.h:24]: is it AbstractPort ptr?

File: src/lava/magma/runtime/cpp_message_infrastructure/include/multiprocessing.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#include "shm.h"
+
+class MultiProcessing {
+ public:
+  // stop each actor in vector actors;
+  void Stop();
+  void BuildActor(std::function<void()>);
+  void CheckActor();
+ private:
+  std::vector<AbstractActor*> actors_;
```
Comment (killight98) [inline src/lava/magma/runtime/cpp_message_infrastructure/include/multiprocessing.h:21]: pls define AbstractActorPtr.

**Review by killight98 — CHANGES_REQUESTED** [review]

**Comment by killight98** [comment]
@hexu33  pls add namespace to each h/cc files as following:
```cpp
namespace XXXX
{

}
```

**Comment by hexu33** [comment]
> @hexu33 pls add namespace to each h/cc files as following:
> 
> ```c++
> namespace XXXX
> {
> 
> }
> ```

Is `msg_passing` OK? Or could you suggest any name else? Do we need sub-namespace?

**Comment by killight98** [comment]
> > @hexu33 pls add namespace to each h/cc files as following:
> > ```c++
> > namespace XXXX
> > {
> > 
> > }
> > ```
> 
> Is `msg_passing` OK? Or could you suggest any name else? Do we need sub-namespace?

or lava_msg_passing?
@hongdami do u have any suggestion?

**Comment by hexu33** [comment]
> @hexu33 pls add namespace to each h/cc files as following:
> 
> ```c++
> namespace XXXX
> {
> 
> }
> ```

File: src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_actor.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+// See: https://spdx.org/licenses/
+
+#ifndef ABSTACT_ACTOR_H_
+#define ABSTACT_ACTOR_H_
+
+namespace message_infrastrature{
```
Comment (killight98) [inline src/lava/magma/runtime/cpp_message_infrastructure/include/abstract_actor.h:8]:
```cpp
namespace message_infrastrature {


}  // namespace message_infrastrature
```

**Review by killight98 — APPROVED** [review]

**Review by hongdami — APPROVED** [review]

**Pull request merged**